### PR TITLE
Add an unique index to service_instance_operations

### PR DESCRIPTION
**Pull request by kathap**
We can see double entries for the same service_instance_id
in the service_instance_operations table. This leads to:
PG::ForeignKeyViolation: ERROR: update or delete on table
"service_instances" violates foreign key constraint
"fk_svc_inst_op_svc_instance_id" on table "service_instance_operations".

Compared to service bindings, service keys and route bindings there is a UNIQUE index
defined for the "resource_id" field; for service instances the UNIQUE keyword is missing. An unique index would prevent to get duplicates, therefore:

* Add unique index on column service_instance_id to prevent double entries for the same
  service_instance
* Remove duplicate entries for one service_instance_id to prepare for adding a uniqueness constraint
* Drop old index before creating the unique one

* Links to any other associated PRs
#2916
#2905

* [X] I have reviewed the [contributing guide](https://github.com/cloudfoundry/cloud_controller_ng/blob/main/CONTRIBUTING.md)

* [X] I have viewed, signed, and submitted the Contributor License Agreement

* [X] I have made this pull request to the `main` branch

* [X] I have run all the unit tests using `bundle exec rake`

* [x] I have run [CF Acceptance Tests](https://github.com/cloudfoundry/cloud_controller_ng#cf-acceptance-tests-cats)

### CODE REVIEW

File: db/migrations/20220818142407_add_unique_index_to_service_instance_operations_service_instance_id.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
+Sequel.migration do
+  up do
+    # Remove duplicate service_instance_operations.service_instance_id to prepare for adding a uniqueness constraint
+    max_updated_at_group = self[:service_instance_operations].
```
Comment (philippthun) [inline db/migrations/20220818142407_add_unique_index_to_service_instance_operations_service_instance_id.rb:4]: I find this hard to understand and thus difficult to verify. So I've pushed a commit with a test for correctly removing duplicates. When executing this test, it fails.

Maybe a different approach would be easier. We don't expect many duplicates, so first finding those service_instance_ids for which there are duplicates and then (for each of them) finding the duplicates that should be removed (keep latest or the one with the highest id) might be easier to implement and understand.

File: db/migrations/20220818142407_add_unique_index_to_service_instance_operations_service_instance_id.rb [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    dup_groups = self[:service_instance_operations].exclude(service_instance_id: nil).
+                 select(:service_instance_id).
+                 group_by(:service_instance_id).
+                 having { count.function.* > 1 }
+
+    dup_groups.each do |group|
+      sorted_ids = self[:service_instance_operations].
+                   select(:id).
+                   where(service_instance_id: group[:service_instance_id]).
+                   map(&:values).
```
Comment (philippthun) [inline db/migrations/20220818142407_add_unique_index_to_service_instance_operations_service_instance_id.rb:14]: `.select(:id).map(&:values)` -> `.select_map(:id)`

File: spec/migrations/20220818142407_add_unique_index_to_service_instance_operations_service_instance_id_spec.rb [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    VCAP::CloudController::ServiceInstanceOperation.instance_exec do
+      @allow_manual_timestamp_update = true
+    end
+
+    # Revert the given migration, i.e. remove the uniqueness constraint.
+    Sequel::Migrator.run(db, tmp_migrations_dir, target: 0, allow_missing_migration_files: true)
+  end
+
+  it 'removes duplicate service instance operations' do
+    # Two operations that do not reference a service instance (i.e. service_instance_id is nil);
```
[the remark this answers is not in the document]
Comment (kathap) [reply spec/migrations/20220818142407_add_unique_index_to_service_instance_operations_service_instance_id_spec.rb:22]: operations with svc_inst_id = nil now don't get deleted at all

File: spec/unit/models/services/service_instance_operation_spec.rb [hunk 4]
```diff
@@ -51,5 +51,37 @@ module VCAP::CloudController
...
+    describe 'when two are created with the same id' do
+      describe 'when a ServiceInstanceOperation exists' do
+        let(:service_instance_attrs) do
+          {
+            name: 'bommel_instance',
+            space: VCAP::CloudController::Space.make
+          }
+        end
+
+        let(:service_instance) { ServiceInstance.create(service_instance_attrs) }
```
Comment (philippthun) [inline spec/unit/models/services/service_instance_operation_spec.rb:64]: You can use `ServiceInstance.make` and remove the `service_instance_attrs`.

File: spec/unit/models/services/service_instance_operation_spec.rb [hunk 5]
```diff
@@ -51,5 +51,37 @@ module VCAP::CloudController
...
+            state: 'in progress',
+            description: '50% all the time',
+            type: 'create',
+            proposed_changes: {
+              name: 'pizza',
+              service_plan_guid: '1800-pizza',
+            },
+          }
+        end
+        before { ServiceInstanceOperation.create(operation_attributes2) }
```
Comment (philippthun) [inline spec/unit/models/services/service_instance_operation_spec.rb:77]: You can use `ServiceInstanceOperation.make(service_instance_id: service_instance.id)` and remove `operation_attributes2`.

**Review by philippthun — APPROVED** [review]

**Pull request merged**